source/opt/loop_peeling.cpp

```diff
@@ -0,0 +1,303 @@
+  // |new_loop| shares the same block as |loop_|.
+  // We mutate all branches form |new_loop| block to |loop_|'s merge into a
+  // branch to |loop_|'s header (so header will also be the merge of
+  // |new_loop|).
+  uint32_t after_loop_pred = 0;
+  for (uint32_t pred_id : cfg.preds(loop_->GetMergeBlock()->id())) {
+    if (loop_->IsInsideLoop(pred_id)) continue;
+    ir::BasicBlock* bb = cfg.block(pred_id);
+    assert(after_loop_pred == 0 && "Predecessor already registered");
+    after_loop_pred = bb->id();
+    bb->ForEachSuccessorLabel([this](uint32_t* succ) {
+      if (*succ == loop_->GetMergeBlock()->id())
+        *succ = loop_->GetHeaderBlock()->id();
+    });
+  }
+
+  // Update cfg.
+  cfg.RemoveNonExistingEdges(loop_->GetMergeBlock()->id());
+  cfg.AddEdge(after_loop_pred, loop_->GetHeaderBlock()->id());
+
+  // Set the merge block of the new loop as the old header block.
+  new_loop_->SetMergeBlock(loop_->GetHeaderBlock());
+
+  // Force the creation of a new preheader and patch the phi of the header.
+  loop_->GetHeaderBlock()->ForEachPhiInst(
+      [after_loop_pred, this](ir::Instruction* phi) {
+        for (uint32_t i = 1; i < phi->NumInOperands(); i += 2) {
+          if (!loop_->IsInsideLoop(phi->GetSingleWordInOperand(i))) {
+            phi->SetInOperand(i, {after_loop_pred});
+            return;
+          }
+        }
+      });
+
+  ConnectIterators(clone_results);
+}
+
+void LoopPeeling::InsertIterator(ir::Instruction* factor) {
+  analysis::Type* factor_type =
+      context_->get_type_mgr()->GetType(factor->type_id());
+  assert(factor_type->kind() == analysis::Type::kInteger);
+  analysis::Integer* int_type = factor_type->AsInteger();
+  assert(int_type->width() == 32);
+
+  InstructionBuilder builder(context_,
+                             &*GetBeforeLoop()->GetLatchBlock()->tail(),
+                             ir::IRContext::kAnalysisDefUse |
+                                 ir::IRContext::kAnalysisInstrToBlockMapping);
+  // Create the increment.
+  // Note that "factor->result_id()" is wrong, the proper id should the phi
+  // value but we don't have it yet. The operand will be set latter, leave
+  // "factor->result_id()" so that the id is a valid and so avoid any assert
+  // that's could be added.
+  ir::Instruction* iv_inc = builder.AddIAdd(
+      factor->type_id(), factor->result_id(),
+      builder.Add32BitConstantInteger<uint32_t>(1, int_type->IsSigned())
+          ->result_id());
+
+  builder.SetInsertPoint(&*GetBeforeLoop()->GetHeaderBlock()->begin());
+
+  extra_induction_variable_ = builder.AddPhi(
+      factor->type_id(),
+      {builder.Add32BitConstantInteger<uint32_t>(0, int_type->IsSigned())
+           ->result_id(),
+       GetBeforeLoop()->GetPreHeaderBlock()->id(), iv_inc->result_id(),
+       GetBeforeLoop()->GetLatchBlock()->id()});
+  // Connect everything.
+  iv_inc->SetInOperand(0, {extra_induction_variable_->result_id()});
```

Comment:
The def-use manager will be out of date after this.  You need to analyze the uses.